File: results/dnd-rest-benchmark/lifecycle-runs/20260731T160058Z_claude_claude-sonnet-5_typescript-vite/src/domain/play/characters.ts
/**
 * Character HP/death, ownership, build, leveling, and skill checks within a
 * play campaign. "Member" (party roster entry) vs. "character owner"
 * (`play_campaign_character_owners`) are distinct: a member is added by
 * joinPlayCampaign and defaults to owning their own character, but ownership
 * can be transferred to another campaign member (see transferCharacter).
 */

import { getDb } from '../../db.ts';
import type { ApiResult, JsonValue } from '../../types.ts';
import { isValidIntInRange } from '../../validation.ts';
import { abilityModifierValue, proficiencyBonusValue } from '../character.ts';
import {
  authenticate,
  isActor,
  isApiResult,
  findCampaign,
  findMemberByCharacterId,
  requireParticipant,
  isMember,
  resolveCharacterOwner,
  requireNonEmptyString,
  type MemberRow,
} from './shared.ts';

export function getCharacterSheet(
  authHeader: string | undefined,
  campaignId: string,
  characterId: string,
): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const forbidden = requireParticipant(db, campaign, actor, 'not a campaign member');
  if (forbidden) return forbidden;

  const owner = resolveCharacterOwner(db, campaignId, characterId);
  if (!owner) {
    return { status: 404, body: { error: 'character not found' } };
  }

  const isDm = campaign.owner === actor.username;
  if (owner !== actor.username && !isDm) {
    return { status: 403, body: { error: 'only the character owner or dm may read this sheet' } };
  }

  const member = findMemberByCharacterId(db, campaignId, characterId);
  if (isApiResult(member)) return member;

  return {
    status: 200,
    body: {
      character_id: characterId,
      owner,
      name: member.name,
      class: member.class,
      level: 1,
      proficiency_bonus: 2,
      hp_max: 10,
      armor_class: 10,
    } as unknown as JsonValue,
  };
}

const OUTCOME_STATUSES = new Set(['success', 'failure']);

export function damageCharacter(
  authHeader: string | undefined,
  campaignId: string,
  characterId: string,
  body: JsonValue,
): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  if (actor.username !== campaign.owner) {
    return { status: 403, body: { error: 'only the owner may apply damage' } };
  }

  const member = findMemberByCharacterId(db, campaignId, characterId);
  if (isApiResult(member)) return member;

  if (!isValidIntInRange(body.amount, 0, 1000000)) {
    return { status: 400, body: { error: 'amount must be a non-negative integer' } };
  }
  const amount = body.amount as number;

  const hpBefore = member.hp_current;
  const hpAfter = Math.max(0, hpBefore - amount);
  let status = member.status;
  let successes = member.death_save_successes;
  let failures = member.death_save_failures;

  if (hpAfter === 0 && member.status === 'conscious') {
    status = 'unconscious';
    successes = 0;
    failures = 0;
  }

  db.prepare(
    'UPDATE play_campaign_members SET hp_current = ?, status = ?, death_save_successes = ?, death_save_failures = ? WHERE campaign_id = ? AND character_id = ?',
  ).run(hpAfter, status, successes, failures, campaignId, characterId);

  return {
    status: 200,
    body: {
      character_id: characterId,
      target: characterId,
      hp_before: hpBefore,
      hp_after: hpAfter,
      hp_current: hpAfter,
      hp_max: member.hp_max,
      status,
      damage: amount,
    },
  };
}

export function recordDeathSave(
  authHeader: string | undefined,
  campaignId: string,
  characterId: string,
  body: JsonValue,
): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const member = findMemberByCharacterId(db, campaignId, characterId);
  if (isApiResult(member)) return member;

  if (actor.username !== member.username) {
    return { status: 403, body: { error: 'only the character owner may roll a death save' } };
  }

  if (member.status !== 'unconscious') {
    return { status: 409, body: { error: 'character is not unconscious' } };
  }

  const outcome = body.outcome;
  if (typeof outcome !== 'string' || !OUTCOME_STATUSES.has(outcome)) {
    return { status: 400, body: { error: 'outcome must be "success" or "failure"' } };
  }

  let successes = member.death_save_successes;
  let failures = member.death_save_failures;
  if (outcome === 'success') {
    successes += 1;
  } else {
    failures += 1;
  }

  let status = member.status;
  if (successes >= 3) {
    status = 'stable';
  } else if (failures >= 3) {
    status = 'dead';
  }

  db.prepare(
    'UPDATE play_campaign_members SET status = ?, death_save_successes = ?, death_save_failures = ? WHERE campaign_id = ? AND character_id = ?',
  ).run(status, successes, failures, campaignId, characterId);

  return {
    status: 201,
    body: { character_id: characterId, successes, failures, status },
  };
}

export function getCharacterOwner(
  authHeader: string | undefined,
  campaignId: string,
  characterId: string,
): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const forbidden = requireParticipant(db, campaign, actor, 'not a campaign member');
  if (forbidden) return forbidden;

  const owner = resolveCharacterOwner(db, campaignId, characterId);
  if (!owner) {
    return { status: 404, body: { error: 'character not found' } };
  }

  return { status: 200, body: { character_id: characterId, owner } };
}

export function claimCharacter(
  authHeader: string | undefined,
  campaignId: string,
  characterId: string,
  _body: JsonValue,
): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const forbidden = requireParticipant(db, campaign, actor, 'not a campaign member');
  if (forbidden) return forbidden;

  const owner = resolveCharacterOwner(db, campaignId, characterId);
  if (owner) {
    return { status: 409, body: { error: 'character is already owned' } };
  }

  db.prepare(
    'INSERT INTO play_campaign_character_owners (campaign_id, character_id, owner) VALUES (?, ?, ?)',
  ).run(campaignId, characterId, actor.username);

  return { status: 201, body: { character_id: characterId, owner: actor.username } };
}

export function transferCharacter(
  authHeader: string | undefined,
  campaignId: string,
  characterId: string,
  body: JsonValue,
): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const forbidden = requireParticipant(db, campaign, actor, 'not a campaign member');
  if (forbidden) return forbidden;

  const owner = resolveCharacterOwner(db, campaignId, characterId);
  if (!owner) {
    return { status: 404, body: { error: 'character not found' } };
  }

  if (owner !== actor.username) {
    return { status: 403, body: { error: 'only the owner may transfer the character' } };
  }

  const newOwner = requireNonEmptyString(body.new_owner, 'new_owner');
  if (isApiResult(newOwner)) return newOwner;

  if (!isMember(db, campaignId, newOwner)) {
    return { status: 400, body: { error: 'new_owner must be a campaign member' } };
  }

  db.prepare(
    'INSERT INTO play_campaign_character_owners (campaign_id, character_id, owner) VALUES (?, ?, ?) ' +
      'ON CONFLICT(campaign_id, character_id) DO UPDATE SET owner = excluded.owner',
  ).run(campaignId, characterId, newOwner);

  return { status: 200, body: { character_id: characterId, owner: newOwner } };
}

const VALID_RACES = new Set([
  'human',
  'elf',
  'dwarf',
  'halfling',
  'gnome',
  'half-elf',
  'half-orc',
  'tiefling',
  'dragonborn',
]);

const VALID_CLASSES = new Set([
  'barbarian',
  'bard',
  'cleric',
  'druid',
  'fighter',
  'monk',
  'paladin',
  'ranger',
  'rogue',
  'sorcerer',
  'warlock',
  'wizard',
]);

const VALID_BACKGROUNDS = new Set([
  'acolyte',
  'charlatan',
  'criminal',
  'entertainer',
  'folk-hero',
  'guild-artisan',
  'hermit',
  'noble',
  'outlander',
  'sage',
  'sailor',
  'soldier',
  'urchin',
]);

const BUILD_ABILITY_KEYS = ['str', 'dex', 'con', 'int', 'wis', 'cha'] as const;

const CLASS_HIT_DIE: Record<string, number> = {
  barbarian: 12,
  bard: 8,
  cleric: 8,
  druid: 8,
  fighter: 10,
  monk: 8,
  paladin: 10,
  ranger: 10,
  rogue: 8,
  sorcerer: 6,
  warlock: 8,
  wizard: 6,
};

export function buildCharacter(
  authHeader: string | undefined,
  campaignId: string,
  characterId: string,
  body: JsonValue,
): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const forbidden = requireParticipant(db, campaign, actor, 'not a campaign member');
  if (forbidden) return forbidden;

  const owner = resolveCharacterOwner(db, campaignId, characterId);
  if (!owner) {
    return { status: 404, body: { error: 'character not found' } };
  }
  if (owner !== actor.username) {
    return { status: 403, body: { error: 'only the character owner may build the character' } };
  }

  const race = body.race;
  if (typeof race !== 'string' || !VALID_RACES.has(race)) {
    return { status: 400, body: { error: 'race must be a valid race' } };
  }

  const characterClass = body.class;
  if (typeof characterClass !== 'string' || !VALID_CLASSES.has(characterClass)) {
    return { status: 400, body: { error: 'class must be a valid class' } };
  }

  const background = body.background;
  if (typeof background !== 'string' || !VALID_BACKGROUNDS.has(background)) {
    return { status: 400, body: { error: 'background must be a valid background' } };
  }

  const abilities = body.abilities;
  if (typeof abilities !== 'object' || abilities === null || Array.isArray(abilities)) {
    return { status: 400, body: { error: 'abilities must be an object' } };
  }
  const abilitiesObj = abilities as JsonValue;

  const scores: Record<string, number> = {};
  for (const key of BUILD_ABILITY_KEYS) {
    const score = abilitiesObj[key];
    if (!isValidIntInRange(score, 1, 30)) {
      return { status: 400, body: { error: `abilities.${key} must be an integer from 1 through 30` } };
    }
    scores[key] = score as number;
  }

  const level = 1;
  const conModifier = abilityModifierValue(scores.con);
  const strModifier = abilityModifierValue(scores.str);
  const dexModifier = abilityModifierValue(scores.dex);
  const intModifier = abilityModifierValue(scores.int);
  const wisModifier = abilityModifierValue(scores.wis);
  const chaModifier = abilityModifierValue(scores.cha);
  const hitDie = CLASS_HIT_DIE[characterClass];
  const hpMax = hitDie + conModifier;
  const proficiencyBonusVal = proficiencyBonusValue(level);

  db.prepare(
    'UPDATE play_campaign_members SET class = ?, hp_current = ?, hp_max = ?, level = ?, con_modifier = ?, str_modifier = ?, dex_modifier = ?, int_modifier = ?, wis_modifier = ?, cha_modifier = ? WHERE campaign_id = ? AND character_id = ?',
  ).run(
    characterClass,
    hpMax,
    hpMax,
    level,
    conModifier,
    strModifier,
    dexModifier,
    intModifier,
    wisModifier,
    chaModifier,
    campaignId,
    characterId,
  );

  return {
    status: 200,
    body: {
      character_id: characterId,
      race,
      class: characterClass,
      background,
      level,
      hp_max: hpMax,
      proficiency_bonus: proficiencyBonusVal,
    },
  };
}

export function levelUpCharacter(
  authHeader: string | undefined,
  campaignId: string,
  characterId: string,
  body: JsonValue,
): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const forbidden = requireParticipant(db, campaign, actor, 'not a campaign member');
  if (forbidden) return forbidden;

  const owner = resolveCharacterOwner(db, campaignId, characterId);
  if (!owner) {
    return { status: 404, body: { error: 'character not found' } };
  }
  if (owner !== actor.username) {
    return { status: 403, body: { error: 'only the character owner may level up the character' } };
  }

  const member = findMemberByCharacterId(db, campaignId, characterId);
  if (isApiResult(member)) return member;

  if (!isValidIntInRange(body.level, 1, 20)) {
    return { status: 400, body: { error: 'level must be an integer from 1 through 20' } };
  }
  const newLevel = body.level as number;
  if (newLevel !== member.level + 1) {
    return { status: 400, body: { error: 'level must be exactly one higher than the current level' } };
  }

  const hitDie = CLASS_HIT_DIE[member.class] ?? 8;
  const perLevelHp = Math.floor(hitDie / 2) + 1 + member.con_modifier;
  const newHpMax = member.hp_max + perLevelHp;
  const proficiencyBonusVal = proficiencyBonusValue(newLevel);

  db.prepare('UPDATE play_campaign_members SET level = ?, hp_max = ? WHERE campaign_id = ? AND character_id = ?').run(
    newLevel,
    newHpMax,
    campaignId,
    characterId,
  );

  return {
    status: 200,
    body: {
      character_id: characterId,
      level: newLevel,
      hp_max: newHpMax,
      hit_dice: `1d${hitDie}`,
      proficiency_bonus: proficiencyBonusVal,
    },
  };
}

const SKILL_ABILITIES: Record<string, string> = {
  acrobatics: 'dex',
  'animal-handling': 'wis',
  arcana: 'int',
  athletics: 'str',
  deception: 'cha',
  history: 'int',
  insight: 'wis',
  intimidation: 'cha',
  investigation: 'int',
  medicine: 'wis',
  nature: 'int',
  perception: 'wis',
  performance: 'cha',
  persuasion: 'cha',
  religion: 'int',
  'sleight-of-hand': 'dex',
  stealth: 'dex',
  survival: 'wis',
};

const ABILITY_MODIFIER_COLUMNS: Record<string, keyof MemberRow> = {
  str: 'str_modifier',
  dex: 'dex_modifier',
  con: 'con_modifier',
  int: 'int_modifier',
  wis: 'wis_modifier',
  cha: 'cha_modifier',
};

export function skillCheck(
  authHeader: string | undefined,
  campaignId: string,
  characterId: string,
  body: JsonValue,
): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const forbidden = requireParticipant(db, campaign, actor, 'not a campaign member');
  if (forbidden) return forbidden;

  const owner = resolveCharacterOwner(db, campaignId, characterId);
  if (!owner) {
    return { status: 404, body: { error: 'character not found' } };
  }
  if (owner !== actor.username) {
    return { status: 403, body: { error: 'only the character owner may make a skill check for the character' } };
  }

  const skill = body.skill;
  if (typeof skill !== 'string' || !(skill in SKILL_ABILITIES)) {
    return { status: 400, body: { error: 'skill must be a supported skill' } };
  }

  const ability = body.ability;
  if (typeof ability !== 'string' || !(ability in ABILITY_MODIFIER_COLUMNS)) {
    return { status: 400, body: { error: 'ability must be a supported ability' } };
  }
  if (SKILL_ABILITIES[skill] !== ability) {
    return { status: 400, body: { error: `skill ${skill} does not use ability ${ability}` } };
  }

  const proficient = body.proficient;
  if (typeof proficient !== 'boolean') {
    return { status: 400, body: { error: 'proficient must be a boolean' } };
  }

  const roll = body.roll;
  if (!isValidIntInRange(roll, 1, 20)) {
    return { status: 400, body: { error: 'roll must be an integer from 1 through 20' } };
  }

  const member = findMemberByCharacterId(db, campaignId, characterId);
  if (isApiResult(member)) return member;

  const abilityModifierVal = member[ABILITY_MODIFIER_COLUMNS[ability]] as number;
  const proficiencyBonusVal = proficient ? proficiencyBonusValue(member.level) : 0;
  const modifier = abilityModifierVal + proficiencyBonusVal;
  const total = (roll as number) + modifier;

  return {
    status: 200,
    body: {
      character_id: characterId,
      skill,
      ability,
      modifier,
      total,
    },
  };
}

export function getCharacterStatus(
  authHeader: string | undefined,
  campaignId: string,
  characterId: string,
): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const forbidden = requireParticipant(db, campaign, actor, 'not a campaign member');
  if (forbidden) return forbidden;

  const member = findMemberByCharacterId(db, campaignId, characterId);
  if (isApiResult(member)) return member;

  return {
    status: 200,
    body: {
      character_id: characterId,
      hp_current: member.hp_current,
      hp_max: member.hp_max,
      status: member.status,
    },
  };
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260804T072004Z_claude_claude-sonnet-5_ruby-sinatra/routes/play/characters.rb
# Per-character play state: HP/death saves outside of an encounter,
# ownership/claim/transfer (a character's controlling player can change
# over a campaign's life), build (initial race/class/ability assignment),
# level-up, and ad-hoc skill checks.

post '/v1/play/campaigns/:id/characters/:char_id/damage' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  require_play_owner!(campaign, user)

  member = find_play_member_by_character!(campaign['id'], params[:char_id])

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  amount = body['amount']
  halt 400, { error: 'invalid amount' }.to_json unless integerish(amount)
  amount = amount.to_i

  hp_before = member['hp_current']
  hp_after = [hp_before - amount, 0].max

  new_status = member['status']
  new_status = 'unconscious' if hp_after.zero? && member['status'] == 'conscious'

  db.execute(
    'UPDATE play_campaign_members SET hp_current = ?, status = ? WHERE campaign_id = ? AND character_id = ?',
    [hp_after, new_status, campaign['id'], params[:char_id]]
  )

  {
    target: params[:char_id],
    hp_before: hp_before,
    hp_after: hp_after,
    damage: amount,
    status: new_status
  }.to_json
end

post '/v1/play/campaigns/:id/characters/:char_id/death-saves' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  member = find_play_member_by_character!(campaign['id'], params[:char_id])

  halt 403, { error: 'forbidden' }.to_json unless member['username'] == user['username']
  halt 409, { error: 'character is not unconscious' }.to_json unless member['status'] == 'unconscious'

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  outcome = body['outcome']
  halt 400, { error: 'invalid outcome' }.to_json unless %w[success failure].include?(outcome)

  successes = member['death_save_successes']
  failures = member['death_save_failures']

  outcome == 'success' ? successes += 1 : failures += 1

  new_status = 'unconscious'
  new_status = 'stable' if successes >= 3
  new_status = 'dead' if failures >= 3

  db.execute(
    'UPDATE play_campaign_members SET death_save_successes = ?, death_save_failures = ?, status = ? ' \
    'WHERE campaign_id = ? AND character_id = ?',
    [successes, failures, new_status, campaign['id'], params[:char_id]]
  )

  status 201
  {
    character_id: params[:char_id],
    successes: successes,
    failures: failures,
    status: new_status
  }.to_json
end

get '/v1/play/campaigns/:id/characters/:char_id/status' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  require_play_participant!(campaign, user, 'not a campaign member')

  member = find_play_member_by_character!(campaign['id'], params[:char_id])

  {
    character_id: params[:char_id],
    hp_current: member['hp_current'],
    hp_max: member['hp_max'],
    status: member['status']
  }.to_json
end

get '/v1/play/campaigns/:id/characters/:char_id/owner' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  require_play_participant!(campaign, user, 'not a campaign member')

  member = find_play_member_by_character!(campaign['id'], params[:char_id])
  owner = play_character_owner(campaign['id'], params[:char_id], member)

  { character_id: params[:char_id], owner: owner }.to_json
end

post '/v1/play/campaigns/:id/characters/:char_id/claim' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  require_play_participant!(campaign, user, 'not a campaign member')

  member = find_play_member_by_character!(campaign['id'], params[:char_id])
  owner = play_character_owner(campaign['id'], params[:char_id], member)

  halt 409, { error: 'character already owned' }.to_json if owner && owner != user['username']

  db.execute(
    'INSERT INTO play_character_owners (campaign_id, character_id, owner) VALUES (?, ?, ?) ' \
    'ON CONFLICT (campaign_id, character_id) DO UPDATE SET owner = excluded.owner',
    [campaign['id'], params[:char_id], user['username']]
  )

  status 201
  { character_id: params[:char_id], owner: user['username'] }.to_json
end

post '/v1/play/campaigns/:id/characters/:char_id/transfer' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  require_play_participant!(campaign, user, 'not a campaign member')

  member = find_play_member_by_character!(campaign['id'], params[:char_id])
  owner = play_character_owner(campaign['id'], params[:char_id], member)

  halt 403, { error: 'not the character owner' }.to_json unless owner == user['username']

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  new_owner = body['new_owner']
  halt 400, { error: 'invalid new_owner' }.to_json unless new_owner.is_a?(String) && !new_owner.empty?
  halt 400, { error: 'new_owner is not a campaign member' }.to_json unless play_campaign_member?(campaign, new_owner)

  db.execute(
    'INSERT INTO play_character_owners (campaign_id, character_id, owner) VALUES (?, ?, ?) ' \
    'ON CONFLICT (campaign_id, character_id) DO UPDATE SET owner = excluded.owner',
    [campaign['id'], params[:char_id], new_owner]
  )

  { character_id: params[:char_id], owner: new_owner }.to_json
end

post '/v1/play/campaigns/:id/characters/:char_id/build' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  require_play_participant!(campaign, user, 'not a campaign member')

  member = find_play_member_by_character!(campaign['id'], params[:char_id])
  owner = play_character_owner(campaign['id'], params[:char_id], member)

  halt 403, { error: 'not the character owner' }.to_json unless owner == user['username']

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  race = body['race']
  klass = body['class']
  background = body['background']
  abilities = body['abilities']

  halt 400, { error: 'invalid race' }.to_json unless VALID_RACES.include?(race)
  halt 400, { error: 'invalid class' }.to_json unless CLASS_HIT_DICE.key?(klass)
  halt 400, { error: 'invalid background' }.to_json unless VALID_BACKGROUNDS.include?(background)
  halt 400, { error: 'invalid abilities' }.to_json unless abilities.is_a?(Hash)

  ABILITY_KEYS.each do |key|
    score = abilities[key]
    halt 400, { error: "invalid ability score #{key}" }.to_json unless integerish(score) && score.to_i.between?(1, 30)
  end

  level = 1
  modifiers = ABILITY_KEYS.each_with_object({}) { |key, acc| acc[key] = ability_modifier(abilities[key].to_i) }
  con_modifier = modifiers['con']
  hp_max = CLASS_HIT_DICE[klass] + con_modifier
  bonus = proficiency_bonus(level)

  db.execute(
    'UPDATE play_campaign_members SET race = ?, class = ?, background = ?, level = ?, hp_max = ?, ' \
    'hp_current = ?, proficiency_bonus = ?, con_modifier = ?, str_modifier = ?, dex_modifier = ?, ' \
    'int_modifier = ?, wis_modifier = ?, cha_modifier = ? WHERE campaign_id = ? AND character_id = ?',
    [race, klass, background, level, hp_max, hp_max, bonus, con_modifier, modifiers['str'], modifiers['dex'],
     modifiers['int'], modifiers['wis'], modifiers['cha'], campaign['id'], params[:char_id]]
  )

  {
    character_id: params[:char_id],
    race: race,
    class: klass,
    background: background,
    level: level,
    hp_max: hp_max,
    proficiency_bonus: bonus
  }.to_json
end

# Fixed (non-random) average hit-point gain for a class's hit die when
# leveling up, per the PHB's "take the average instead of rolling" option:
# floor(die / 2) + 1. All hit dice in CLASS_HIT_DICE are even, so this is
# exact (e.g. d8 -> 5, d10 -> 6).
def average_hit_die_gain(klass)
  (CLASS_HIT_DICE[klass] / 2) + 1
end

post '/v1/play/campaigns/:id/characters/:char_id/level-up' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  require_play_participant!(campaign, user, 'not a campaign member')

  member = find_play_member_by_character!(campaign['id'], params[:char_id])
  owner = play_character_owner(campaign['id'], params[:char_id], member)

  halt 403, { error: 'not the character owner' }.to_json unless owner == user['username']

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  new_level = body['level']
  halt 400, { error: 'invalid level' }.to_json unless integerish(new_level)

  current_level = member['level']
  halt 400, { error: 'level must be exactly one higher than the current level' }.to_json unless new_level.to_i == current_level + 1

  klass = member['class']
  halt 400, { error: 'invalid class' }.to_json unless CLASS_HIT_DICE.key?(klass)

  con_modifier = member['con_modifier']
  hp_gain = average_hit_die_gain(klass) + con_modifier
  hp_max = member['hp_max'] + hp_gain
  hp_current = [member['hp_current'] + hp_gain, hp_max].min
  bonus = proficiency_bonus(new_level.to_i)

  db.execute(
    'UPDATE play_campaign_members SET level = ?, hp_max = ?, hp_current = ?, proficiency_bonus = ? ' \
    'WHERE campaign_id = ? AND character_id = ?',
    [new_level.to_i, hp_max, hp_current, bonus, campaign['id'], params[:char_id]]
  )

  {
    character_id: params[:char_id],
    level: new_level.to_i,
    hp_max: hp_max,
    hit_dice: "1d#{CLASS_HIT_DICE[klass]}",
    proficiency_bonus: bonus
  }.to_json
end

get '/v1/play/campaigns/:id/characters/:char_id/sheet' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  is_owner = require_play_participant!(campaign, user, 'not a campaign member')

  member = find_play_member_by_character!(campaign['id'], params[:char_id])
  owner = play_character_owner(campaign['id'], params[:char_id], member)

  halt 403, { error: 'forbidden' }.to_json unless is_owner || owner == user['username']

  {
    character_id: params[:char_id],
    owner: owner,
    name: member['name'],
    class: member['class'],
    level: 1,
    proficiency_bonus: 2,
    hp_max: 10,
    armor_class: 10
  }.to_json
end

post '/v1/play/campaigns/:id/characters/:char_id/skill-check' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  require_play_participant!(campaign, user, 'not a campaign member')

  member = find_play_member_by_character!(campaign['id'], params[:char_id])
  owner = play_character_owner(campaign['id'], params[:char_id], member)

  halt 403, { error: 'not the character owner' }.to_json unless owner == user['username']

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  skill = body['skill']
  ability = body['ability']
  proficient = body['proficient']
  roll = body['roll']

  halt 400, { error: 'unsupported skill' }.to_json unless SKILL_ABILITIES.key?(skill)
  halt 400, { error: 'unsupported ability' }.to_json unless ability == SKILL_ABILITIES[skill]
  halt 400, { error: 'invalid proficient' }.to_json unless [true, false].include?(proficient)
  halt 400, { error: 'invalid roll' }.to_json unless integerish(roll)

  ability_mod = member["#{ability}_modifier"]
  modifier = ability_mod + (proficient ? member['proficiency_bonus'] : 0)
  total = roll.to_i + modifier

  {
    character_id: params[:char_id],
    skill: skill,
    ability: ability,
    modifier: modifier,
    total: total
  }.to_json
end
